File: airline/backend/app/schemas/booking.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from app.models.booking import PaymentMethod, BookingStatus
from app.schemas.flight import Flight
from app.schemas.user import UserProfile


class BookingBase(BaseModel):
    flight_id: int
    seat_number: str


class BookingCreate(BookingBase):
    payment_method: PaymentMethod


class Booking(BookingBase):
    id: int
    pnr: Optional[str] = None
    passenger_id: int
    price: float
    payment_method: Optional[PaymentMethod] = None
    status: BookingStatus
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    passport_number: Optional[str] = None
    date_of_birth: Optional[datetime] = None
    created_at: datetime
    confirmed_at: Optional[datetime] = None
    flight: Flight
    
    # Extra fields for staff view
    checked_in: bool = False
    qr_code: Optional[str] = None

    class Config:
        from_attributes = True

    @classmethod
    def model_validate(cls, obj, **kwargs):
        # Handle SQLAlchemy model
        instance = super().model_validate(obj, **kwargs)
        if hasattr(obj, 'ticket') and obj.ticket:
            instance.checked_in = obj.ticket.checked_in
            instance.qr_code = obj.ticket.qr_code
        return instance


class SeatConflict(BaseModel):
    seat_number: str
    bookings: List[Booking]


class SeatHoldRequest(BaseModel):
    flight_id: int
    seat_number: str


class SeatHoldResponse(BaseModel):
    id: int
    flight_id: int
    seat_number: str
    expires_at: datetime

    class Config:
        from_attributes = True


class Ticket(BaseModel):
    id: int
    booking_id: int
    passenger_id: int
    flight_id: int
    seat_number: str
    qr_code: Optional[str] = None
    checked_in: bool
    checked_in_at: Optional[datetime] = None
    flight: Flight
    created_at: datetime

    class Config:
        from_attributes = True


class CheckInRequest(BaseModel):
    ticket_id: int


class CheckInResponse(BaseModel):
    ticket: Ticket
    boarding_pass: str  # QR code string

File: airline/flutter_app/lib/models/booking.dart
import 'flight.dart';

enum PaymentMethod {
  card,
  applePay,
  googlePay;

  static PaymentMethod fromString(String method) {
    switch (method.toUpperCase()) {
      case 'CARD':
        return PaymentMethod.card;
      case 'APPLE_PAY':
        return PaymentMethod.applePay;
      case 'GOOGLE_PAY':
        return PaymentMethod.googlePay;
      default:
        return PaymentMethod.card;
    }
  }

  String get displayName {
    switch (this) {
      case PaymentMethod.card:
        return 'Банковская карта';
      case PaymentMethod.applePay:
        return 'Apple Pay';
      case PaymentMethod.googlePay:
        return 'Google Pay';
    }
  }
}

enum BookingStatus {
  pending,
  confirmed,
  cancelled;

  static BookingStatus fromString(String status) {
    switch (status.toUpperCase()) {
      case 'PENDING':
        return BookingStatus.pending;
      case 'CONFIRMED':
        return BookingStatus.confirmed;
      case 'CANCELLED':
        return BookingStatus.cancelled;
      default:
        return BookingStatus.pending;
    }
  }
}

class Booking {
  final int id;
  final String? pnr; // Added PNR
  final int passengerId;
  final int flightId;
  final Flight flight;
  final String seatNumber;
  final double price;
  final PaymentMethod? paymentMethod;
  final BookingStatus status;
  final String? firstName;
  final String? lastName;
  final String? passportNumber;
  final DateTime? dateOfBirth;
  final DateTime createdAt;
  final DateTime? confirmedAt;
  final DateTime? expiresAt;
  final bool checkedIn;
  final String? qrCode;

  Booking({
    required this.id,
    this.pnr,
    required this.passengerId,
    required this.flightId,
    required this.flight,
    required this.seatNumber,
    required this.price,
    this.paymentMethod,
    required this.status,
    this.firstName,
    this.lastName,
    this.passportNumber,
    this.dateOfBirth,
    required this.createdAt,
    this.confirmedAt,
    this.expiresAt,
    this.checkedIn = false,
    this.qrCode,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] as int? ?? 0,
      pnr: json['pnr'] as String?,
      passengerId: json['passenger_id'] as int? ?? 0,
      flightId: json['flight_id'] as int? ?? 0,
      flight: Flight.fromJson(json['flight'] as Map<String, dynamic>? ?? {}),
      seatNumber: (json['seat_number'] ?? 'N/A').toString(),
      price: (json['price'] as num? ?? 0).toDouble(),
      paymentMethod: json['payment_method'] != null
          ? PaymentMethod.fromString(json['payment_method'].toString())
          : null,
      status: BookingStatus.fromString((json['status'] ?? 'pending').toString()),
      firstName: json['first_name'] as String?,
      lastName: json['last_name'] as String?,
      passportNumber: json['passport_number'] as String?,
      dateOfBirth: _parseDate(json['date_of_birth']),
      createdAt: _parseDate(json['created_at']) ?? DateTime.now(),
      confirmedAt: _parseDate(json['confirmed_at']),
      expiresAt: _parseDate(json['expires_at']),
      checkedIn: json['checked_in'] as bool? ?? false,
      qrCode: json['qr_code'] as String?,
    );
  }

  static DateTime? _parseDate(dynamic date) {
    if (date == null) return null;
    try {
      if (date is String && date.isNotEmpty) {
        return DateTime.parse(date).toLocal();
      }
      return null;
    } catch (_) {
      return null;
    }
  }
}



